### PR TITLE
Merge branch dev_zip into dev (1)

### PR DESCRIPTION
**Pull request by younesdessia**
* **What kind of changes does this PR introduce?** 
- [ ] Bug fix
- [X] new features
- [ ] performance 
- [ ] docs update

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
- [ ] Yes:
- [X] No

* **Other information**:
this feature allows you to export in zip format from the object page

image (9)

### CODE REVIEW

File: dessia_common/core.py [hunk 1]
```diff
@@ -722,13 +723,39 @@ def save_babylonjs_to_file(self, filename: str = None, use_cdn: bool = True, deb
...
+                        f"Archive input is not a file-like object. Got '{value}' of type {type(value)}")
+        return [archive]
+
     def _export_formats(self) -> List[ExportFormat]:
         """ Return a list of objects describing how to call 3D exports. """
         formats = DessiaObject._export_formats(self)
         formats3d = [ExportFormat(selector="step", extension="step", method_name="to_step_stream", text=True),
                      ExportFormat(selector="stl", extension="stl", method_name="to_stl_stream", text=False),
                      ExportFormat(selector="html", extension="html", method_name="to_html_stream", text=True)]
+        export_zip = ExportFormat(selector="zip", extension="zip", method_name="to_zip_stream", text=False)
```
Comment (GhislainJ) [inline dessia_common/core.py:756]: Could simply add this to the format3d list

File: dessia_common/core.py [hunk 2]
```diff
@@ -722,13 +723,39 @@ def save_babylonjs_to_file(self, filename: str = None, use_cdn: bool = True, deb
...
+        step_stream = dcf.StringFile("export_step.stp")
+        self.to_step_stream(step_stream)
+        html_stream = self.to_html_stream(dcf.StringFile(filename="export_html.html"))
+        stl_stream = dcf.BinaryFile("export_stl.stl")
+        self.to_stl_stream(stl_stream)
+
+        list_stream = [step_stream, html_stream, stl_stream]
+        archive_name = 'export_zip'
+        archive.filename = archive_name
+        with ZipFile(archive, 'w') as zip_archive:
```
Comment (GhislainJ) [inline dessia_common/core.py:737]: This is duplicated from workflow Archive blocks and need to be mutualized. Consider adding a function to files.py module

**Review by GhislainJ — CHANGES_REQUESTED** [review]
How do we ensure genericity ? What if someone doesn't want to add HMTL to its zip ?

File: dessia_common/core.py (hunk 1 above)
Comment (younesdessia) [reply dessia_common/core.py:756]: Yes I can, I just wanted to separate the two

File: dessia_common/core.py [hunk 3]
```diff
@@ -607,10 +608,50 @@ def to_xlsx_stream(self, stream):
...
+
+        json_stream = dcf.JsonFile("json_export.json")
+        self.save_to_stream(json_stream)
+
+        list_stream = [excel_stream, json_stream]
+        return list_stream
+
+    def to_zip_stream(self, archive: dcf.BinaryFile) -> List[dcf.BinaryFile]:
+        """ Creates a zip archive containing several files representing the export of DessiaObject. """
+        list_stream = self.zip_settings()
```
Comment (GhislainJ) [inline dessia_common/core.py:624]: Avoid setting the type of the variable in its name. dc's conventions state that we should name lists with an "s" at the end.

Here, for example `streams`

File: dessia_common/core.py [hunk 4]
```diff
@@ -607,10 +608,50 @@ def to_xlsx_stream(self, stream):
...
+        archive_name = 'export_zip'
+        archive.filename = archive_name
+        with ZipFile(archive, 'w') as zip_archive:
+            for value in list_stream:
```
Comment (GhislainJ) [inline dessia_common/core.py:628]: Following previous comment : `for stream in streams`

File: dessia_common/core.py [hunk 5]
```diff
@@ -607,10 +608,50 @@ def to_xlsx_stream(self, stream):
...
+                if isinstance(value, dcf.StringFile):
+                    with zip_archive.open(value.filename, 'w') as file:
+                        file.write(value.getvalue().encode('utf-8'))
+                elif isinstance(value, dcf.BinaryFile):
+                    with zip_archive.open(value.filename, 'w') as file:
+                        file.write(value.getbuffer())
+                else:
+                    raise ValueError(
```
Comment (GhislainJ) [inline dessia_common/core.py:636]: Could probably be set on one line

File: dessia_common/core.py [hunk 6]
```diff
@@ -607,10 +608,50 @@ def to_xlsx_stream(self, stream):
...
 
+    def zip_settings(self):
+        """ Returns a list of streams containing different representations of the object. """
+        excel_stream = dcf.BinaryFile("excel_export.xlsx")
+        self.to_xlsx_stream(excel_stream)
+
+        json_stream = dcf.JsonFile("json_export.json")
+        self.save_to_stream(json_stream)
+
+        list_stream = [excel_stream, json_stream]
```
Comment (GhislainJ) [inline dessia_common/core.py:619]: could directly return variable content instead of assigning `list_stream` and returning it

**Review by GhislainJ — CHANGES_REQUESTED** [review]

**Comment by younesdessia** [comment]
@GhislainJ it's good for modifications

File: dessia_common/core.py [hunk 7]
```diff
@@ -605,10 +607,41 @@ def to_xlsx_stream(self, stream):
...
+        streams = self.zip_settings()
+        archive_name = 'export_zip'
+        archive.filename = archive_name
+        with ZipFile(archive, 'w', compression=zipfile.ZIP_DEFLATED) as zip_archive:
+            for value in streams:
+                archive = dcf.generate_archive(zip_archive=zip_archive, value=value)
+        return [archive]
+
+    def to_zip(self, filepath: str):
+        """ Creates a zip archive containing several files representing the export of DessiaObject. """
```
[the remark this answers is not in the document]
Comment (younesdessia) [reply dessia_common/core.py]: it's ok

File: dessia_common/core.py [hunk 8]
```diff
@@ -605,10 +607,41 @@ def to_xlsx_stream(self, stream):
         writer = XLSXWriter(self)
         writer.save_to_stream(stream)
 
+    def zip_settings(self):
+        """ Returns a list of streams containing different representations of the object. """
```
[the remark this answers is not in the document]
Comment (younesdessia) [reply dessia_common/core.py]: it's ok

File: code_pylint.py [hunk 9]
```diff
@@ -26,8 +26,8 @@
 UNWATCHED_ERRORS = ["fixme", "trailing-whitespace", "import-error"]
 
 MAX_ERROR_BY_TYPE = {
-    "wrong-spelling-in-docstring": 164,
-    "wrong-spelling-in-comment": 13,
+    "wrong-spelling-in-docstring": 179,
```
[the remark this answers is not in the document]
Comment (younesdessia) [reply code_pylint.py]: it's ok, I fixed it

**Review by GhislainJ — APPROVED** [review]

**Pull request merged**